File: Day284/sumOfXOROfAllPairs.go
package main

import (
	"fmt"
)

func sumXOR(arr []int, n int) int64 {
	var answer int64
	for i := 0; i < 32; i++ {
		var temp1, temp2 int64
		for j := 0; j < n; j++ {
			if (arr[j] & 1) == 0 {
				temp1++
			} else {
				temp2++
			}
			arr[j] >>= 1
		}
		answer += temp1 * temp2 * (1 << i)
	}
	return answer
}

func main() {
	var testcases int
	fmt.Scan(&testcases)

	for testcases > 0 {
		var sizeOfArray int
		fmt.Scan(&sizeOfArray)

		arr := make([]int, sizeOfArray)
		for i := 0; i < sizeOfArray; i++ {
			fmt.Scan(&arr[i])
		}

		res := sumXOR(arr, sizeOfArray)
		fmt.Println(res)
		testcases--
	}
}

File: Day284/sumOfXOROfAllPairs.java
//{ Driver Code Starts
//Initial Template for Java

//Initial Template for Java

//Initial Template for Java

/*package whatever //do not write package name here */

import java.io.*;
import java.util.*;


class Array {
    
    // Driver code
	public static void main (String[] args) throws IOException{
		// Taking input using buffered reader
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		int testcases = Integer.parseInt(br.readLine());
		
		// looping through all testcases
		while(testcases-- > 0){
		    String line = br.readLine();
		    String[] element = line.trim().split("\\s+");
		    int sizeOfArray = Integer.parseInt(element[0]);
		    
		    int arr [] = new int[sizeOfArray];
		    
		    line = br.readLine();
		    String[] elements = line.trim().split("\\s+");
		    for(int i = 0;i<sizeOfArray;i++){
		        arr[i] = Integer.parseInt(elements[i]);
		    }
		    
		    
		    Solution obj = new Solution();
		    long res = obj.sumXOR(arr, sizeOfArray);
		    System.out.println(res);
		}
	}
}



// } Driver Code Ends


//User function Template for Java

class Solution{
   
    // Function for finding maximum and value pair
    public long sumXOR (int arr[], int n) {
        //Complete the function
        long answer = 0;
        for (int i=0; i<32; i++) {
            long temp1 = 0, temp2 = 0;
            for (int j=0; j<n; j++) {
                if ((arr[j]&1)==0) {
                    temp1++;
                } else {
                    temp2++;
                }
                arr[j]>>=1;
            }
            answer += temp1 * temp2 *(1<<i);
        }
        return answer;
    }
    
    
}


